crates/typst-macros/src/symbols.rs

```diff
@@ -29,19 +33,37 @@ pub fn symbols(stream: TokenStream) -> Result<TokenStream> {
     Ok(quote! { &[#(#pairs),*] })
 }
 
+fn construct_sym_char(ch: &LitChar, handler: &Handler) -> TokenStream {
+    match handler {
+        Handler::None => quote! { ::typst::symbols::SymChar::pure(#ch), },
+        Handler::Path(path) => quote! {
+            ::typst::symbols::SymChar::with_func(
+                #ch,
+                <#path as ::typst::foundations::NativeFunc>::func,
+            ),
+        },
+    }
+}
+
 struct Symbol {
     name: syn::Ident,
     kind: Kind,
 }
 
 enum Kind {
-    Single(syn::LitChar),
+    Single(syn::LitChar, Handler),
     Multiple(Punctuated<Variant, Token![,]>),
 }
 
 struct Variant {
     name: String,
     c: syn::LitChar,
+    handler: Handler,
+}
+
+enum Handler {
+    None,
+    Path(Path),
 }
```

Comment:
In my experience, redefined `Option` is a bit of an anti-pattern because you don't get the option methods with it (e.g. `is_none` which you had to redefine). You could try having a newtype `struct Handler(Option<Path>)` instead and see whether it ends up nicer or not.